plutus-contract/test/Spec/Contract/Tx/Constraints/MustUseOutputAsCollateral.hs

```diff
@@ -0,0 +1,326 @@
+import Plutus.ChainIndex.Emulator (diskState, unCredentialMap)
+import Plutus.ChainIndex.Emulator.DiskState (addressMap)
+import Plutus.Contract as Con
+import Plutus.Contract.Test (assertUnbalancedTx, assertValidatedTransactionCount,
+                             assertValidatedTransactionCountOfTotal, checkPredicateOptions, defaultCheckOptions,
+                             emulatorConfig, mockWalletPaymentPubKeyHash, w1, w2, (.&&.))
+import Plutus.Script.Utils.Typed (Any)
+import Plutus.Script.Utils.V1.Scripts qualified as PSU.V1
+import Plutus.Script.Utils.V2.Scripts qualified as PSU.V2
+import Plutus.Script.Utils.V2.Typed.Scripts qualified as V2.Scripts
+import Plutus.Trace.Emulator qualified as Trace (EmulatorTrace, activateContractWallet, nextSlot, params,
+                                                 setSigningProcess, walletInstanceTag)
+import Plutus.V1.Ledger.Api (Address (addressCredential))
+import Plutus.V1.Ledger.Value qualified as Value
+import PlutusTx qualified
+import PlutusTx.Prelude qualified as P
+import Wallet.Emulator.Wallet as Wallet (WalletState, chainIndexEmulatorState, ownAddress, signPrivateKeys,
+                                         walletToMockWallet')
+
+tests :: TestTree
+tests =
+    testGroup "MustUseOutputAsCollateral"
+      [ testGroup "ledger constraints" $ [v1Tests, v2Tests] ?? ledgerSubmitTx
+      --, testGroup "cardano constraints" $ [v1Tests, v2Tests] ?? cardanoSubmitTx
+      ]
+
+v1Tests :: SubmitTx -> TestTree
+v1Tests sub = testGroup "Plutus V1" $
+   [ v1FeaturesTests
+   --, v2FeaturesNotAvailableTests
+   ] ?? sub ?? PSU.PlutusV1
+
+v2Tests :: SubmitTx -> TestTree
+v2Tests sub = testGroup "Plutus V2" $
+  [ v1FeaturesTests
+  --, v2FeaturesTests
+  ] ?? sub ?? PSU.PlutusV2
+
+v1FeaturesTests :: SubmitTx -> PSU.Language -> TestTree
+v1FeaturesTests sub t = testGroup "Plutus V1 features" $
+    [ singleUseOfMustUseOutputAsCollateral
+    , multipleUseOfMustUseOutputAsCollateral
+    , usingMustUseOutputAsCollateralWithOtherWalletUtxo
+    , usingMustUseOutputAsCollateralWithScriptUtxo
+    , useOfMustUseOutputAsCollateralWithoutPlutusScript
+    , ledgerValidationErrorWhenMustUseOutputAsCollateralExceedsMaximumCollateralInputs
+    -- (no phase2 tests because this error can never occur)
+    -- (no tests to check balanced tx for when balancer does/doesn't include collateral inputs)
+    ] ?? sub ?? t
+
+utxoValue :: Value.Value
+utxoValue = Ada.lovelaceValueOf 10_000_000
+
+tknAmount :: Integer
+tknAmount = 5_000_000
+
+tknValueOf :: Integer -> PSU.Language -> Value.Value
+tknValueOf x tc = Value.singleton (mustUseOutputAsCollateralPolicyCurrencySymbol tc) "mint-me" x
+
+tknValue :: PSU.Language -> Value.Value
+tknValue = tknValueOf tknAmount
+
+w1PaymentPubKeyHash :: L.PaymentPubKeyHash
+w1PaymentPubKeyHash = mockWalletPaymentPubKeyHash w1
+
+w2PaymentPubKeyHash :: L.PaymentPubKeyHash
+w2PaymentPubKeyHash = mockWalletPaymentPubKeyHash w2
+
+maximumCollateralInputs :: Integer
```

Comment:
I wouldn't hardcode this value because it's part of the protocol parameters. Two options:

* Use the `Def` instance of `ProtocolParameters` in `Ledger.Params` right here. It's the `protocolParamMaxCollateralInputs` parameter.
* Or instead query the PP in the `Contract` Monad directly, but using `getNodeParams`, and get the `protocolParamMaxCollateralInputs` from there.